src/Products/PluginIndexes/KeywordIndex/KeywordIndex.py

```diff
@@ -58,59 +72,120 @@ def _index_object(self, documentId, obj, threshold=None, attr=''):
-            rdiff = difference(newKeywords, oldKeywords)
+            if oldKeywords in (missing, empty):
+                self.removeSpecialIndexEntry(oldKeywords, documentId)
+                oldSet = OOSet()
+            else:
+                if not isinstance(oldKeywords, OOSet):
+                    oldKeywords = OOSet(oldKeywords)
+                oldSet = oldKeywords
+
+            if newKeywords in (missing, empty):
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+                newSet = OOSet()
+            else:
+                newSet = newKeywords = OOSet(newKeywords)
+
+            fdiff = difference(oldSet, newSet)
+            rdiff = difference(newSet, oldSet)
             if fdiff or rdiff:
                 # if we've got forward or reverse changes
-                if newKeywords:
-                    self._unindex[documentId] = list(newKeywords)
-                else:
-                    del self._unindex[documentId]
                 if fdiff:
                     self.unindex_objectKeywords(documentId, fdiff)
                 if rdiff:
                     for kw in rdiff:
                         self.insertForwardIndexEntry(kw, documentId)
+
+            self._unindex[documentId] = newKeywords
+
         return 1
 
     def _get_object_keywords(self, obj, attr):
-        newKeywords = getattr(obj, attr, ())
+        newKeywords = getattr(obj, attr, None)
+
+        def _getSpecialValueFor(datum):
+            try:
+                special_value = self.special_values[datum]
+            except TypeError:
+                raise KeyError(datum)
+
+            if self.providesSpecialIndex(special_value):
+                return special_value
+            raise KeyError(datum)
+
         if safe_callable(newKeywords):
             try:
                 newKeywords = newKeywords()
             except (AttributeError, TypeError):
-                return ()
-        if not newKeywords:
-            return ()
-        elif isinstance(newKeywords, basestring):
-            return (newKeywords,)
+                LOG.debug('%(context)s: Cannot determine datum for attribute '
+                          '%(attr)s of object %(obj)r', dict(
+                              context=self.__class__.__name__,
+                              attr=attr,
+                              obj=obj),
+                          exc_info=True)
+
+                newKeywords = sys.exc_info()[0]
+                try:
+                    return _getSpecialValueFor(newKeywords)
+                except KeyError:
+                    return _marker
+
+        try:
+            return _getSpecialValueFor(newKeywords)
+        except KeyError:
+            pass
+
+        # normalize datum
+        if isinstance(newKeywords, basestring):
+            newKeywords = (newKeywords,)
         else:
             try:
                 # unique
                 newKeywords = set(newKeywords)
```

Comment:
At another place, the keywords are collected in an `OOSet`. Using different set types increases the constraints placed on the usable types for keywords: `OOSet` requires orderability (as the `BTrees` package as a whole); `set` requires hashability. I recommend to use `OOSet` uniformly (and avoid the `tuple` recasting).

Reply:
Okay